cybersecurity/offensive/c2/apollo.yml

```diff
@@ -0,0 +1,746 @@
+        type: string
+        description: The hostname to download the file from.
+    cmdline:
+      - download
+      - -Path
+      - ${Path}
+      - -Host
+      - ${Host}
+  execute_assembly:
+    description: Execute a .NET assembly registered with register_file.
+    parameters:
+      assembly:
+        type: string
+        description: The assembly to execute.
+      args:
+        type: string
+        description: Arguments to pass to the assembly.
+    cmdline:
+      - execute_assembly
+      - -Assembly
+      - ${assembly}
+      - -Arguments
+      - ${args}
+  execute_coff:
+    description: Execute a object file (BOF) that's been registered with register_file.
+    parameters:
+      object:
+        type: string
+        description: The object file to execute.
+      function:
+        type: string
+        description: The function to execute.
+      timeout:
+        type: integer
+        description: The timeout for the execution.
+      args:
+        type: string
+        description: Arguments to pass to the object file.
+    cmdline:
+      - execute_coff
+      - -Coff
+      - ${object}
+      - -Function
+      - ${function}
+      - -Timeout
+      - ${timeout}
+      - -Arguments
+      - ${args}
+  execute_pe:
+    description: Execute a statically compiled executable that's been registered with register_file.
+    parameters:
+      binary:
+        type: string
+        description: The binary to execute.
+      args:
+        type: string
+        description: Arguments to pass to the binary.
+    cmdline:
+      - execute_pe
+      - -PE
+      - ${binary}
+      - -Arguments
+      - ${args}
+  exit:
+    description: Task agent to exit.
+    cmdline:
+      - exit
+  get_injection_techniques:
+    description: Show currently registered injection techniques as well as the current technique.
+    cmdline:
+      - get_injection_techniques
+  get_privs:
+    description: Enable as many privileges as possible for the current access token.
+    cmdline:
+      - getprivs   
+  ifconfig:
+    description: Get Network Adapters and Interfaces
+    cmdline:
+      - ifconfig
+  inject:
+    description: Inject a new payload into a remote process.
+    parameters:
+      pid:
+        type: integer
+        description: The process ID to inject into.
+      payload:
+        type: string
+        description: The payload to inject.
+    cmdline:
+      - inject
+      - -PID
+      - ${pid}
+      - -Payload
+      - ${payload}
+  inline_assembly:
+    description: Execute a .NET assembly in the currently executing process that's been registered with register_file.
+    parameters:
+      Assembly:
+        type: string
+        description: The assembly to execute.
+      Arguments:
+        type: string
+        description: Additional arguments to pass to the assembly.
+    cmdline:
+      - inline_assembly
+      - -Assembly
+      - ${Assembly}
+      - -Arguments
+      - ${Arguments}
+  jobkill: 
```

Comment:
all these blocks seem auto generated ... i'm not sure i understand how these are intended to work tho, most don't seem like commonly available shell commands, but specific commands for apollo, while robopages execute their tools via a standard shell (wether on host, in a container or via ssh somewhere else)

Reply:
Maybe we could add a new type of execution adapter that would execute a specific yml by interacting with a custom target process/shell?

If i understand the intention here all these commands are supposed to do Apollo what to do, effectively using an LLM agent as its C2, so it could make sense to have what we have right now:

* Shell execution - default behaviour, executes cmdline as a shell command on the host)
* SSH execution - same as above but executing on a different host 
* Docker execution - runs the cmdline inside the specified container image (for tools that are not installed by default on the host and/or to limit what the agent can do)

Plus a new:

* Custom execution adapter - that would cover these cases (another I can think of is msfconsole for instance, or bettercap).

So before we can merge this we have to answer:

### 1. How would the code sending these commands to Apollo look like?

Especially as part of this namespace https://github.com/dreadnode/robopages-cli/tree/main/src/runtime

### 2. How can we generalize that?

So that we don't have to implemen dozens of specific interactions with things like Apollo but just one that works for all (as an objective, if we can generalize to both Apollo and msfconsole that'd be great). Maybe an integrated plugin system and/or scripting engine? I'd like to avoid this complexity unless strictly needed, like if we see that we can't generalize the code easily.

### 3. How do we differentiate with normal robopages cmdlines (that execute via shell)?

How do we define in the YML file that we need this custom adapter rather than a standard shell? Maybe something like:

```yml
adapter:
    type: stdin # writes the cmdline to the stdin of a target process
    target: /path/to/msfconsole # or the apollo_executable?
```

or 

```yml
adapter:
    type: rest # sends the cmdline as a json request to an API
    target: https://example.com:8080/v1/api/something
```

or 

```yml
adapter:
    type: tcp # sends the cmdline as a tcp buffer to and endpoint
    target: tcp-server-ip:1234
```

etc etc